src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -2999,23 +2999,53 @@ private void VisitStatementsWithLocalFunctions(BoundBlock block)
             if (!TrackingRegions && !block.LocalFunctions.IsDefaultOrEmpty)
             {
                 // First visit everything else
+                var queue1 = ArrayBuilder<BoundLocalFunctionStatement>.GetInstance();
                 foreach (var stmt in block.Statements)
                 {
-                    if (stmt.Kind != BoundKind.LocalFunctionStatement)
+                    if (stmt is BoundLocalFunctionStatement localFunc)
+                    {
+                        queue1.Add(localFunc);
+                    }
+                    else
                     {
                         VisitStatement(stmt);
                     }
                 }
 
-                // Now visit the local function bodies
-                foreach (var stmt in block.Statements)
+                // Now visit the local function bodies.
+                // We first visit those we have seen called (hence we know their starting state),
+                // repeating this while seeing more call sites inside other local functions,
+                // and only then we visit the completely unreachable bodies.
+                // This avoids unnecessary passes and incorrect starting states.
+                var queue2 = ArrayBuilder<BoundLocalFunctionStatement>.GetInstance();
+                for (int c = 0; c != LocalFuncUsageCount;)
                 {
-                    if (stmt is BoundLocalFunctionStatement localFunc)
+                    c = LocalFuncUsageCount;
```

Comment:
>c = LocalFuncUsageCount; [](http://example.com/codeflow?start=20&length=24)

It doesn't look like ```c``` is changed after this point, this means that the enclosing loop will not continue, and, effectively, we will make at most one iteration. Why do we need the loop then?

Reply:
> effectively, we will make at most one iteration

I realized that this is actually not accurate. The ```LocalFuncUsageCount``` can be incremented by analysis for one of the reachable local function.